### PR TITLE
230 form validation

### PR DESCRIPTION
**Pull request by carltonsmith**
### This pull request changes...

This PR adds form validation to the `EditProfile` page.

**FUNCTIONALITY**
There should be no validation done until the user tries to submit the form. Upon submission, any form field that does not have a valid character (one character for first and last name and a selected STT) will receive a form group error (the line on the side of the form group), a red outline will surround the corresponding input missing its value, and a statement will appear to explain what is missing above the element.

**TO TEST**
- Go to https://tdp-frontend.app.cloud.gov/ and login
- On the login page, focus on the First Name input and then blur away from the input
    - There should be no error yet
- Repeat previous step for Last Name
    - There should be no error yet
- Repeat step for STT
    - There should be no error yet
- Click on the `Request Access` button
- All 3 inputs should have the above mentioned styling changes
- Go to `First Name` input, type in a character and blur out of it and the error messaging should go away
- Repeat for `Last Name`
- Repeat for `STT`

EDIT: Previously the card noted an outstanding issue  [370](https://github.com/raft-tech/TANF-app/issues/370). That issue has now been cleared.

### [AC] This pull request is ready to merge when...

- [x] Tests have been updated (and all tests are passing)
- [x] Vulnerability Scan Test (Zap) has no high/critical CVEs
- [x] Linting Tests (ESLint for UI and Flake8 for Python)
- [x] Unit Tests ( including a code coverage test report)
- [x] Accessibility Tests (Pa11y UI Only)
- [x] This code has been reviewed by someone other than the original author
- [x] The experience passes a basic manual accessibility audit (keyboard nav, screen reader, text scaling) OR an exemption is documented
- [x] The change has been documented
- [x] Associated OpenAPI documentation has been updated
- [x] Changelog is updated as appropriate

### [DoD] This feature is done when...

- [x] Design has approved the experience
- [x] Product has approved the experience

### CODE REVIEW

**Review by alexsoble — CHANGES_REQUESTED** [review]
As part of QASP review, I ran the Accessibility Insights Fast Pass automated check against the error state with no First or Last name entered.

Accessibility Insights found the following error:

**Error**
duplicate-id: Ensures every id attribute value is unique

**How to fix**
Fix the following: Document has multiple static elements with the same id attribute: input-error-message

Visual helper for this error:

Screen Shot 2020-10-29 at 2 20 29 PM

File: tdrs-frontend/src/components/EditProfile/EditProfile.jsx [hunk 1]
```diff
@@ -1,65 +1,222 @@
...
+        >
+          <label
+            className={`usa-label ${errors.lastName ? 'usa-label--error' : ''}`}
+            htmlFor="lastName"
+          >
+            Last name
+            {errors.lastName && (
+              <span
+                className="usa-error-message"
+                id="input-error-message"
```
Comment (alexsoble) [inline tdrs-frontend/src/components/EditProfile/EditProfile.jsx:168]: @hilvitzs We'll want different ids for the different error messages.

File: tdrs-frontend/src/components/ComboBox/ComboBox.jsx [hunk 2]
```diff
@@ -1,21 +1,72 @@
...
   return (
-    <label className="usa-label" htmlFor="sttList">
+    <label className="usa-label" htmlFor={name.toUpperCase()}>
       Associated State, Tribe, or Territory
-      <div className="usa-combo-box" data-placeholder="- Select or Search -">
-        <select className="usa-select" name="stt" id="stt" ref={selectRef}>
+      {error && (
+        <span
+          className="usa-error-message"
+          id="input-error-message"
```
Comment (alexsoble) [inline tdrs-frontend/src/components/ComboBox/ComboBox.jsx:52]: @hilvitzs This looks like another location where we may be introducing the same id for different elements, which we'll want to resolve.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,7 +1,8 @@
 version: 2.1
 jobs:
   test-backend:
-    machine: true
+    machine:
```
Comment (alexsoble) [inline .circleci/config.yml:4]: Noting that these CircleCI config changes are incidental to this PR and were already merged in via #69.

File: tdrs-frontend/src/actions/stts.js [hunk 4]
```diff
@@ -5,13 +5,38 @@ export const SET_STTS = 'SET_STTS'
 export const SET_STTS_ERROR = 'SET_STTS_ERROR'
 export const CLEAR_STTS = 'CLEAR_STTS'
 
+/**
+ * This action fires an HTTP GET request to the API
```
Comment (alexsoble) [inline tdrs-frontend/src/actions/stts.js:9]: This is great inline documentation!

File: tdrs-frontend/src/components/ComboBox/ComboBox.jsx [hunk 5]
```diff
@@ -1,21 +1,72 @@
-import React, { useEffect, useRef } from 'react'
+import React, { useEffect } from 'react'
 import PropTypes from 'prop-types'
 import comboBox from 'uswds/src/js/components/combo-box'
 
-const ComboBox = ({ children }) => {
-  const selectRef = useRef()
+/**
+ * @param {Component(s)} children - One or more React components to be rendered
```
Comment (alexsoble) [inline tdrs-frontend/src/components/ComboBox/ComboBox.jsx:6]: This is great inline documentation! x2

File: tdrs-frontend/src/components/EditProfile/EditProfile.jsx (hunk 1 above)
Comment (hilvitzs) [reply tdrs-frontend/src/components/EditProfile/EditProfile.jsx]: Thanks for pointing this out. I pushed the fix for both this and the other spot in combo box where this was happening.

Comment (alexsoble) [reply tdrs-frontend/src/components/EditProfile/EditProfile.jsx]: Excellent! Is the fix deployed to dev?

Comment (hilvitzs) [reply tdrs-frontend/src/components/EditProfile/EditProfile.jsx]: Yes it is

Comment (alexsoble) [reply tdrs-frontend/src/components/EditProfile/EditProfile.jsx]: Thanks!

**Comment by alexsoble** [comment]
Thanks for fixing that issue @hilvitzs! The Accessibility Insights Fast Pass automated check now passes. 🎉 

Next steps: @lfrohlich and I plan to review ACs together tomorrow.

@iamjolly and I plan to run the full a11y assessment tomorrow on this PR as well and record the session; @ttran-hub can't make it unfortunately.

**Comment by hilvitzs** [comment]
@alexsoble No problem! Thanks for taking a look. I'll look forward to hearing about the accessibility checks.

**Comment by alexsoble** [comment]
Just noting that the ACs for this issue are here: https://github.com/raft-tech/TANF-app/issues/230.

**Review by lfrohlich — APPROVED** [review]
Check all ACs - this looks really good! Once the a11y review is done I am good to merge.

**Comment by iamjolly** [comment]
Accessibility Review notes:

1. **Change requested:** Fields that are required should also have a visual indication in the label, something like `First Name (required)`.
2. **Feedback:** The visual format of the inline error/validation messages is good. 💯 
3. **Change requested:** When the form is submitted with multiple errors, only the last error is announced to screen reader users (Safari + VoiceOver). This may be due to setting multiple `span`s with `role="alert"` at the same time. Instead, it would be best to display a new `div` or `span` containing the error summary (something like "There were X errors with your form submission.") and place that above the First Name `label` on the form. 
4. **Change requested:** When the form is submitted with error(s), the focus remains on the `button` element. I recommend setting focus to the error summary element (use `tabindex="-1"` to allow programmatic focus) if there are errors. That way, a screen reader user will not need to move backward through the form to find the errors and correct them.
5. **Reminder:** The `label ... for="STT"...` remains as a mismatch with the `input ... id="stt" ...` element which should be resolved. This was discussed as part of the USWDS/React combo-box issue previously.

**Comment by hilvitzs** [comment]
Hi @iamjolly. Thanks so much for taking a look! I'll work on getting those fixes in for Monday.

**Comment by alexsoble** [comment]
@iamjolly New work has been pushed to the dev site; in addition to seeing the visual indication of "(required)", I just reviewed with VoiceOver and found that when the form is submitted with errors:

1. I hear an error summary ("There are 3 errors in this form.")
2. Focus shifts away from the submit button and onto the error summary element. 

@iamjolly When you have a moment, would you mind taking a look and letting us know if this passes a11y review or if more changes are needed? Thanks!

**Comment by iamjolly** [comment]
@hilvitzs and @alexsoble - I just tested this with Safari + VoiceOver, and it looks good to me and resolves all of the requested changes for accessibility. Thanks!

**Comment by alexsoble** [comment]
### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

- [ ] [Error message is displayed in a way consistent with USWDS](https://designsystem.digital.gov/components/form-controls/#validation) when fields are not valid
- [ ] Testing Checklist has been run and all tests pass
- [ ] New issues for form timeouts is created

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ [x] Frontend code changes tested in `ComboBox.test.js`.
+ [x] Frontend test coverage: 100% (see https://github.com/raft-tech/TANF-app README coverage dashboard).

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Fontend code style checks passing on CircleCI.

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

+ [x] All a11y issues surfaced during reviews with @iamjolly and @ttran-hub using Accessibility Insights addressed.

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*

+ [x] Code successfully deployed via automated CircleCI process to development on Cloud.gov.

### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] Frontend code accompanied by thorough inline documentation.

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] OWASP Scan passes on CircleCI.

**Comment by alexsoble** [comment]
@lfrohlich All the non-AC QASP elements are passing, but when I visit the dev site I'm not seeing the latest changes — for example, I still see "We need to collect some information before an OFA Admin can grant you access" rather than the latest copy. Are others seeing the same thing when they visit the dev site?

**Comment by alexsoble** [comment]
@lfrohlich Alrighty, now we're both seeing the latest on dev!

**Review by alexsoble — APPROVED** [review]

**Pull request merged**